### PR TITLE
fix: resolve critical lifetime, overflow, and concurrency bugs across core modules

### PR DESCRIPTION
**Pull request by tkhquang**
## Summary

Resolve all critical and high-severity issues identified in the v2.0.0 release readiness audit.

### Critical Fixes
- **C1**: StringPool missing placement-new — add constructed_mask tracking and proper destruction
- **C2/C3**: Logger use-after-free — change async_logger_ from unique_ptr to shared_ptr with lock-safe copy
- **C4**: Integer overflow in is_entry_valid_and_covers — add overflow guards for address + size arithmetic

### High-Severity Fixes
- **H1**: Add RAII ActiveReaderGuard for s_activeReaders — eliminates counter drift on early returns
- **H2**: Fast-path atomic check on async_mode_enabled_ before acquiring lock
- **H3**: Fix pending_messages_ counter drift — use memory_order_acq_rel, fix DropOldest underflow
- **H4**: Reduce block_timeout_ms default from 100ms to 16ms (one frame at 60fps)
- **H6**: Reduce MAX_FOLLOWER_YIELDS from 64 to 8 for faster VirtualQuery fallback

### Medium Fixes
- **M1**: Config default_value initialized from moved-from value — reorder init list to copy before move
- **M8**: Stale msg_size after truncation — move capture to after resize()

### Tests
- 14 new regression tests covering all fixed code paths (447 total, all passing)
- Fix OverflowPolicy_Block test to use explicit block_timeout_ms matching the new 16ms default

### Docs
- Update README block timeout description (100ms → 16ms)

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved async-logger enqueue/overflow and flush robustness; fixed counter/race issues and addressed message truncation behavior.
  * Hardened address-range checks to prevent arithmetic wraparound misreports.

* **Performance**
  * Reduced follower wait times and improved reader-tracking for render-critical paths.

* **Changes**
  * Default async logger block timeout reduced from 100ms to 16ms (one frame at 60fps).

* **Documentation**
  * README updated to reflect the new async logger timeout.

* **Tests**
  * Expanded tests for async logger, config defaults, and memory edge cases.

### CODE REVIEW

File: src/memory.cpp [hunk 1]
```diff
@@ -948,14 +977,11 @@ void DetourModKit::Memory::invalidate_range(const void *address, size_t size)
     if (!MemoryUtilsCacheInternal::s_cacheInitialized.load(std::memory_order_acquire))
         return;
 
-    MemoryUtilsCacheInternal::s_activeReaders.fetch_add(1, std::memory_order_acq_rel);
+    MemoryUtilsCacheInternal::ActiveReaderGuard reader_guard;
```
[the remark this answers is not in the document]
Comment (tkhquang) [reply src/memory.cpp:980]: The race window between reading s_cacheInitialized and constructing ActiveReaderGuard is real but benign. The safety comes from the second check: shutdown_cache() zeroes s_shardCount (line 877) *before* entering the drain loop (line 882). Any reader that passes the s_cacheInitialized check but races with shutdown will:

1. Construct ActiveReaderGuard (incrementing s_activeReaders)
2. Read s_shardCount == 0
3. Return immediately without touching any data structures
4. Destruct the guard (decrementing s_activeReaders), unblocking shutdown

The data structures (s_cacheShards, s_shardMutexes) are only destroyed after the drain loop completes (lines 888-901), so no reader can access freed memory. The s_shardCount == 0 early-return is the actual admission gate, not s_cacheInitialized — the latter is a fast-path optimization to avoid the atomic increment on the common "cache not initialized" path.

Adding a formal admission protocol would add an atomic CAS to every is_readable/is_writable call for a race that already has a safe fallback. Not worth the hot-path cost.

File: tests/test_memory.cpp [hunk 2]
```diff
@@ -1177,3 +1174,64 @@ TEST_F(MemoryTest, CacheHitRate_RepeatedAccess)
...
+    }
+
+    // Shutdown should wait for readers to finish without crash
+    Memory::shutdown_cache();
+    reader.join();
+
+    EXPECT_TRUE(reader_done.load());
+
+    // Re-init for TearDown
+    Memory::init_cache();
```
[the remark this answers is not in the document]
Comment (tkhquang) [reply tests/test_memory.cpp:1240]: This test's purpose is to verify shutdown_cache() doesn't crash or deadlock when concurrent readers exist — it's a safety regression test for the H1 fix (ActiveReaderGuard), not a deterministic race test.

Signaling reader_started only after the first is_readable() call would require either instrumenting production code or adding test-only hooks into the ActiveReaderGuard path, which we avoid for a test-only concern. The 100-iteration read loop already provides a high probability of overlap with shutdown_cache(). If we needed deterministic race testing, we'd use ThreadSanitizer (which we do run in CI) rather than hand-rolled signaling.

The test as written catches the original bug (unbalanced s_activeReaders causing shutdown to hang) and any crash from use-after-free during concurrent shutdown — both validated by the fact that it completes within the test timeout.

File: src/async_logger.cpp [hunk 3]
```diff
@@ -403,7 +419,7 @@ namespace DetourModKit
 
         if (queue_.try_push(msg))
         {
-            pending_messages_.fetch_add(1, std::memory_order_relaxed);
+            pending_messages_.fetch_add(1, std::memory_order_acq_rel);
             flush_cv_.notify_one();
```
[the remark this answers is not in the document]
Comment (tkhquang) [reply src/async_logger.cpp:426]: The scenario described — log() snapshots async_logger_ via shared_ptr copy, then shutdown() joins the writer thread while enqueue() is in-flight — is safe by design. This is exactly why we changed from unique_ptr to shared_ptr (the C2/C3 fix):

1. log() copies the shared_ptr under async_mutex_ (line 220), giving it a strong reference
2. log() releases async_mutex_ and calls enqueue() on the local copy (line 224)
3. shutdown() acquires async_mutex_, moves async_logger_ out, and calls AsyncLogger::shutdown() (lines 171-179)
4. AsyncLogger::shutdown() sets its running flag to false, joins the writer thread, and returns
5. The in-flight enqueue() is still safe because:
   - The shared_ptr in log() keeps the AsyncLogger alive
   - enqueue() pushes to a lock-free queue — it doesn't require the writer thread to be running
   - The message will simply remain in the queue (not drained), which is acceptable during shutdown

The shared_ptr prevents use-after-free. The worst case is a message enqueued after the writer thread joins — it stays in the queue and is destroyed with the AsyncLogger when the last shared_ptr drops. No crash, no UB. Adding an in-flight counter would serialize the shutdown path on every log() call for a scenario that's already safe.

[the remark this answers is not in the document]
Comment (tkhquang) [reply src/async_logger.cpp:426]: > `@tkhquang`, thank you for the detailed explanation — you are entirely correct that the `shared_ptr` ownership change keeps the `AsyncLogger` alive through any in-flight `enqueue()` call, and no use-after-free can occur. That concern is well-handled by the C2/C3 fix.
> 
> However, my original comment was about a different, narrower race — specifically the ordering between `try_push` and `pending_messages_.fetch_add`:
> 
> ```
> Thread A (enqueue):                   Thread B (flush):
>   queue_.try_push(msg)  ← succeeds
>   // ← preempted HERE
>                                         pending_messages_ == 0 → TRUE
>                                         flush_with_timeout returns "flushed"
>   pending_messages_.fetch_add(1, ...)
>   flush_cv_.notify_one()
> ```
> 
> At the moment Thread B observes `pending_messages_ == 0`, a message is already sitting in the lock-free queue but hasn't been counted yet. The stronger memory order (`acq_rel`) improves cross-thread visibility of the counter itself, but it doesn't close this specific window because the push and the counter increment are **two separate operations** with a preemption point between them.
> 
> Concretely: a caller that calls `flush()` immediately after `log()` from a **different** thread may get a false "all flushed" signal.
> 
> If this race is acceptable as a known limitation of the flush contract (i.e., `flush()` provides a best-effort rather than strict guarantee), that's a reasonable design trade-off and I won't press further. But it would be worth documenting this in a code comment near `flush_with_timeout` so future maintainers don't assume flush is strictly ordered with respect to concurrent `enqueue()` calls.
> 
> `(-.-)` _noted_
> ✏️ Learnings added
> 
> ```
> Learnt from: tkhquang
> Repo: tkhquang/DetourModKit PR: 17
> File: src/async_logger.cpp:420-423
> Timestamp: 2026-03-21T19:22:53.070Z
> Learning: In `src/async_logger.cpp` (DetourModKit), `AsyncLogger::enqueue()` calls `queue_.try_push(msg)` BEFORE `pending_messages_.fetch_add(1, ...)`. This creates a narrow window where a concurrent `flush_with_timeout()` can observe `pending_messages_ == 0` and return early while a message is already in the lock-free queue but not yet counted. This is a known limitation of the flush contract: flush() provides best-effort rather than strict ordering guarantees with respect to concurrent enqueue() calls. The `memory_order_acq_rel` change improves visibility but does not eliminate this TOCTOU window.
> ```
> 
> > Note: Learnings are effective only in the context of similar code segments. To apply general coding guidelines, please configure [review instructions](https://docs.coderabbit.ai/guides/review-instructions). You can manage existing learnings in the [app](https://app.coderabbit.ai/learnings).
> 
> 🧠 Learnings used
> 
> ```
> Learnt from: tkhquang
> Repo: tkhquang/DetourModKit PR: 17
> File: src/memory.cpp:977-980
> Timestamp: 2026-03-21T19:16:04.830Z
> Learning: In `src/memory.cpp` (DetourModKit), the admission-gate for accessing `s_cacheShards`/`s_shardMutexes` is `s_shardCount == 0`, NOT `s_cacheInitialized`. `shutdown_cache()` zeroes `s_shardCount` BEFORE entering the `s_activeReaders` drain loop, and data structures are only destroyed after the drain loop completes. The `s_cacheInitialized` fast-path check is a hot-path optimization to skip the `s_activeReaders` atomic increment when the cache is not initialized. Any reader that races past the `s_cacheInitialized` check will increment `s_activeReaders` (holding up shutdown's drain loop) and then see `s_shardCount == 0`, returning immediately without touching any data structures. This design is intentional and safe.
> ```
> 
> ```
> Learnt from: tkhquang
> Repo: tkhquang/DetourModKit PR: 17
> File: tests/test_memory.cpp:1197-1223
> Timestamp: 2026-03-21T19:15:57.896Z
> Learning: In the DetourModKit repository (tests/test_memory.cpp), concurrent-safety regression tests (e.g., ShutdownCache_ConcurrentReaders) intentionally signal reader_started before entering Memory::is_readable() rather than after, to avoid instrumenting production code. The 100-iteration read loop is used for probabilistic overlap with shutdown_cache(). Deterministic race coverage is delegated to ThreadSanitizer running in CI. This is intentional design — do not flag the early flag-set as a correctness issue.
> ```

Good catch. The race between try_push and pending_messages_.fetch_add is real:
```
Thread A (enqueue):                   Thread B (flush):
  queue_.try_push(msg)  ← succeeds
  // ← preempted HERE
                                        pending_messages_ == 0 → TRUE
                                        flush_with_timeout returns "flushed"
  pending_messages_.fetch_add(1, ...)
```

Fixed by moving the increment before the push in both affected paths:

1. Main enqueue path (L420): pending_messages_ is incremented before try_push. On push failure, it's decremented before entering handle_overflow.

2. Block overflow path (L618): pending_messages_ is incremented once before the retry loop. On timeout, it's decremented before returning false. This ensures flush sees the in-flight message throughout the entire blocking retry window.

The DropOldest path was already correct — it replaces an already-counted message (pop then push), so the net count stays balanced and there's no window where an uncounted message sits in the queue.

Cost: one extra atomic decrement on the push-failure cold path. No impact on the hot path (successful push).

File: src/async_logger.cpp [hunk 4]
```diff
@@ -129,7 +144,8 @@ namespace DetourModKit
             {
                 auto offset = static_cast<size_t>(raw_ptr - block_begin);
                 PoolSlot *slot = reinterpret_cast<PoolSlot *>(b->data) + (offset / sizeof(PoolSlot));
-                slot->str.~basic_string();
+                slot->str.clear();
+                slot->str.shrink_to_fit();
```
[the remark this answers is not in the document]
Comment (tkhquang) [reply src/async_logger.cpp:149]: `std::string::shrink_to_fit()` does not throw in practice. The C++ standard specifies it as a non-binding request with no Throws clause — unlike `reserve()`, it has no `length_error` precondition. All major implementations (libstdc++, libc++, MSVC STL) implement it as noexcept or equivalent: on failure to reallocate smaller, the request is simply ignored (the non-binding escape hatch).

The `noexcept` guarantee on `deallocate()` is not violated.

That said, if the concern is about unnecessary work on the deallocation hot path: `shrink_to_fit()` is intentional here. Without it, a slot that once held a large message retains its heap allocation across reuse cycles, causing the pool's steady-state memory to drift toward the high-water mark of the largest message ever seen. The `shrink_to_fit()` call reclaims that capacity so recycled slots return to SSO size. This is a deliberate memory-vs-CPU tradeoff for a logging pool that may see occasional large messages mixed with many small ones.
